custom/enikshay/duplicate_ids.py

```diff
@@ -0,0 +1,96 @@
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from dimagi.utils.couch.database import iter_docs
+
+from .const import VOUCHER_ID, ENIKSHAY_ID
+
+
+def get_cases_with_duplicate_ids(domain, case_type, all_case_ids):
+    accessor = CaseAccessors(domain)
+    id_property = {'voucher': VOUCHER_ID, 'person': ENIKSHAY_ID}[case_type]
+    all_cases = [
+        {
+            'case_id': case.case_id,
+            'readable_id': case.get_case_property(id_property),
+            'opened_on': case.opened_on,
+        }
+        for case in accessor.iter_cases(all_case_ids)
+    ]
+    counts = Counter(case['readable_id'] for case in all_cases)
+    bad_cases = (case for case in all_cases if counts[case['readable_id']] > 1)
+    return sorted(bad_cases, key=lambda case: case['opened_on'], reverse=True)
+
+
+def get_bad_case_info(domain, case_type, full_debug_info=False):
+    accessor = CaseAccessors(domain)
+    case_ids = accessor.get_case_ids_in_domain(case_type)
+    bad_cases = get_cases_with_duplicate_ids(domain, case_type, case_ids)
+    add_debug_info_to_cases(bad_cases, full_debug_info)
+    context = {
+        'case_type': case_type,
+        'num_bad_cases': len(bad_cases),
+        'num_total_cases': len(case_ids),
+        'num_good_cases': len(case_ids) - len(bad_cases),
+        'bad_cases': bad_cases,
+    }
+    return context
+
+
+def add_debug_info_to_cases(bad_cases, full_debug_info):
+    _add_form_info_to_cases(bad_cases, full_debug_info)
+    _add_user_info_to_cases(bad_cases)
+
+
+def _add_form_info_to_cases(bad_cases, full_debug_info):
+    for case in bad_cases[:None if full_debug_info else 300]:
+        form = CaseAccessorSQL.get_transactions(case['case_id'])[0].form
+        if form:
+            case['form_name'] = form.form_data.get('@name', 'NA')
+            form_device_number = form.form_data.get('serial_id', {}).get('outputs', {}).get('device_number')
+            case['device_number_in_form'] = form_device_number
+            case['form_device_id'] = form.metadata.deviceID
+            case['form_user_id'] = form.user_id
+            case['auth_user_id'] = form.auth_context.get('user_id')
+
+
+def _add_user_info_to_cases(bad_cases):
+    user_info = _get_user_info(
+        case['form_user_id'] for case in bad_cases if 'form_user_id' in case)
+
+    auth_user_ids = [case['auth_user_id'] for case in bad_cases
+                     if 'auth_user_id' in case]
+    auth_usernames = {user_doc['_id']: user_doc['username'] for user_doc in
+                      iter_docs(CommCareUser.get_db(), auth_user_ids)}
+    for case in bad_cases:
+        user_dict = user_info.get(case.get('form_user_id'))
+        if user_dict:
+            case['username'] = user_dict['username']
+            device_id = case['form_device_id']
+            if device_id == 'Formplayer':
+                if case['form_user_id'] == case['auth_user_id']:
+                    device_id = "WebAppsLogin"
```

Comment:
Can this be declared as a constant somewhere?

Reply:
I'm not sure that's appropriate.  This is a hack that I wouldn't want to grant that much legitimacy.  WebApps uses different device IDs in restores and in form submissions, and this code takes a guess as to what WebApps probably used in the restore.  I added a comment about it.